### PR TITLE
OSDOCS#1491 - Add IPsec support

### PR DESCRIPTION
**Pull request by jboxman**
- https://issues.redhat.com/browse/OSDOCS-1491

Previews:

- [Installing a cluster on AWS with network customizations](https://deploy-preview-27911--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-network-customizations.html#nw-operator-configuration-parameters-for-ovn-sdn_installing-aws-network-customizations)
- [IPsec encryption configuration](https://deploy-preview-27911--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/about-ipsec-ovn.html)
- [About the OVN-Kubernetes default Container Network Interface (CNI) network provider](https://deploy-preview-27911--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/about-ovn-kubernetes.html#nw-ovn-kubernetes-matrix_about-ovn-kubernetes)
- [About the OpenShift SDN default CNI network provider](https://deploy-preview-27911--osdocs.netlify.app/openshift-enterprise/latest/networking/openshift_sdn/about-openshift-sdn.html#nw-ovn-kubernetes-matrix_about-openshift-sdn)
- [Cluster Network Operator in OpenShift Container Platform](https://deploy-preview-27911--osdocs.netlify.app/openshift-enterprise/latest/networking/cluster-network-operator.html#nw-operator-configuration-parameters-for-ovn-sdn_cluster-network-operator)

### CODE REVIEW

**Comment by jboxman** [comment]
Hi @markdgray,

Can you review this documentation for IPsec for inclusion in the OCP docs?

- Is any important information missing?
- Is any information incorrect?

I'm not particularly familiar with IPsec and encryption mechanics in general; Are there any concepts or details that should be elaborated on for clarity?

Thanks!

**Comment by jboxman** [comment]
Hi @anuragthehatter,

This is the PR to add IPsec documentation for OCP.

What do you think?

Thanks!

**Comment by jboxman** [comment]
@anuragthehatter, ping.

File: networking/ovn_kubernetes_network_provider/about-ipsec-ovn.adoc [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+toc::[]
+
+With IPsec enabled, all network traffic between pods on the OVN-Kubernetes Container Network Interface (CNI) cluster network travels through an encrypted tunnel.
+
+IPsec is disabled by default.
+
+[NOTE]
+====
+IPsec encryption can be enabled only during cluster installation.
```
Comment (markdgray) [inline networking/ovn_kubernetes_network_provider/about-ipsec-ovn.adoc:14]: Also it can't be disabled. Worth calling out.

Comment (anuragthehatter) [reply networking/ovn_kubernetes_network_provider/about-ipsec-ovn.adoc:14]: @markdgray Need to mention about modes like transport?

File: modules/nw-ovn-ipsec-traffic.adoc [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Module included in the following assemblies:
+//
+
+[id="nw-ovn-ipsec-traffic_{context}"]
+= Types of network traffic flows encrypted by IPsec
+
+With IPsec enabled, only the following network traffic flows between pods are encrypted:
+
```
Comment (anuragthehatter) [inline modules/nw-ovn-ipsec-traffic.adoc:8]: We might need to re-visit this section. As pr testing following is encrypted or not:
Left side is traffic originator and right side is receiver as follow
```
Hostnw-pod -> pod is encrypted
pod -> pod is encrypted
pod-> hotnw-pod in NOT
Hostnw-pod ->Hostnw-pod is NOT
```

File: networking/ovn_kubernetes_network_provider/about-ipsec-ovn.adoc (hunk 1 above)
Comment (markdgray) [reply networking/ovn_kubernetes_network_provider/about-ipsec-ovn.adoc:14]: @anuragthehatter Yes, it is worth calling out.

File: modules/nw-ovn-ipsec-traffic.adoc (hunk 2 above)
Comment (markdgray) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: Yeah, I think you are right. I even read it and through it was ok. Maybe a table like what @anuragthehatter has would be clearer.

Comment (jboxman) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @markdgray, I'm going to put in a request for a diagram and include that, with the enablement slides as a template for that. (There are certain requirements for public facing assets, so they're created from scratch.)

Comment (markdgray) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @jboxman Another thing that we should probably include is that this feature is for single stack IPv4 clusters only

Comment (jboxman) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @markdgray, will that be true when dual stack is available?

Comment (jboxman) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @markdgray @anuragthehatter 

And to clarify my understanding, given:

- _Pod A_, A pod on the host network
- _Pod Z_, A pod on the cluster network

If Pod A send an HTTP request to to Pod Z, the HTTP request would traverse the IPsec tunnel, but the response would not traverse the tunnel? (So the traffic is encrypted unidirectionally? Which is to say also that the IP traffic follows a different route, inbound and outbound on the network infrastructure?)

Comment (markdgray) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: > @markdgray, will that be true when dual stack is available?

Yes, we will explicitly enable this

Comment (markdgray) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: > And to clarify my understanding, given:
> 
>     * _Pod A_, A pod on the host network
> 
>     * _Pod Z_, A pod on the cluster network
> 
> 
> If Pod A send an HTTP request to to Pod Z, the HTTP request would traverse the IPsec tunnel, but the response would not traverse the tunnel? (So the traffic is encrypted unidirectionally? Which is to say also that the IP traffic follows a different route, inbound and outbound on the network infrastructure?)

No, what is important is who initiates the connection. Pod A->Pod Z will get SNAT'ed on the way out of the node so an return communications would go over the pod network and get encrypted.

Comment (jboxman) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @anuragthehatter, does it look okay now?

Thanks!

Comment (anuragthehatter) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @jboxman Could you guide me the right path on netlify to read that? I tried to find under networking but couldn't

Comment (jboxman) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: @anuragthehatter should be: https://deploy-preview-27911--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/about-ipsec-ovn.html

Comment (anuragthehatter) [reply modules/nw-ovn-ipsec-traffic.adoc:8]: Thanks Jason. That looks good to me.

File: modules/nw-ovn-ipsec-certificates.adoc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// Module included in the following assemblies:
+//
+
+[id="nw-ovn-ipsec-certificates_{context}"]
+= Security certificate generation and rotation
+
+The Cluster Network Operator (CNO) generates a self-signed X.509 certificate authority (CA) that is used by IPsec for encryption. Certificate signing requests (CSRs) from each node are automatically fulfilled by the CNO.
+
+The CA is valid for 10 years. The individual node certificates are valid for 5 years and are automatically rotated after 4 and a half years elapse.
```
Comment (bobfuru) [inline modules/nw-ovn-ipsec-certificates.adoc:9]: I wouldn't spell out 4-and-a-half years. 
"after 4 1/2 years elapse" is a more standardized convention.

File: modules/nw-ovn-ipsec-encryption.adoc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Module included in the following assemblies:
+//
+
+[id="nw-ovn-ipsec-encryption_{context}"]
+= Encryption protocol and tunnel mode for IPsec
+
+The encrypt cipher used is `AES-GCM-16-256`. The integrity check value (ICV) is `16` bytes. The key length is `256` bits.
+
+The IPsec tunnel mode used is _Transport mode_, a mode that encryptions end-to-end communication.
```
Comment (bobfuru) [inline modules/nw-ovn-ipsec-encryption.adoc:9]: s/encryptions/encrypts

**Review by bobfuru — COMMENTED** [review]
A couple of small things and otherwise LGTM!

File: modules/nw-ovn-ipsec-certificates.adoc (hunk 3 above)
Comment (jboxman) [reply modules/nw-ovn-ipsec-certificates.adoc:9]: @bobfuru, so you're saying I'm nonstandard now? :)

**Comment by jboxman** [comment]
@bobfuru thanks!

**Pull request merged**